### PR TITLE
feat(virtualization): add stop,start,restart commands

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
Added stop,start, restart subcommands for `d8 virtualization`

### CODE REVIEW

File: internal/virtualization/cmd/powerstate/vmop/vmop.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+func (v VirtualMachineOperation) generateMsg(vmop *v1alpha2.VirtualMachineOperation) string {
+	if vmop == nil {
+		return ""
+	}
+	key := types.NamespacedName{Namespace: vmop.GetNamespace(), Name: vmop.GetName()}
+	phase := vmop.Status.Phase
+	if !v.isCompleted(vmop) {
+		return fmt.Sprintf("VirtualMachineOperation %q did not complete. phase=%q\n", key.String(), phase)
+	}
+	return fmt.Sprintf("VirtualMachineOperation %q compleated. phase=%q, reason=%q, message=%q\n", key.String(), phase, vmop.Status.FailureReason, vmop.Status.FailureMessage)
```
Comment (diafour) [inline internal/virtualization/cmd/powerstate/vmop/vmop.go:68]:
```suggestion
	return fmt.Sprintf("VirtualMachineOperation %q completed. phase=%q, reason=%q, message=%q\n", key.String(), phase, vmop.Status.FailureReason, vmop.Status.FailureMessage)
```

Comment (diafour) [inline internal/virtualization/cmd/powerstate/vmop/vmop.go:68]: These messages lacks operation type, i.e. start, stop.

```
VirtualMachineOperation "my-super-important-host-12eqw" completed. phase=Completed, reason=, message=
```
This is not informative, it is better to print operation type:

```
VirtualMachine "my-super-important-host" started. VirtualMachineOperation "my-super-important-host-12eqw" completed. 
```

Also, it seems "completed" message should not print phase, reason, and message, but "not completed" should print reason and message.

File: internal/virtualization/cmd/powerstate/vmop/vmop.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		return nil, context.DeadlineExceeded
+	}
+	return vmop, nil
+}
+
+func (v VirtualMachineOperation) isCompleted(vmop *v1alpha2.VirtualMachineOperation) bool {
+	if vmop == nil {
+		return false
+	}
+	return vmop.Status.Phase == v1alpha2.VMOPPhaseCompleted || vmop.Status.Phase == v1alpha2.VMOPPhaseFailed
```
Comment (diafour) [inline internal/virtualization/cmd/powerstate/vmop/vmop.go:117]: Oh, failed is completed. It gives misleading messages in generateMsg.

File: internal/virtualization/cmd/powerstate/powerstate.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+		Force:   false,
+		Wait:    false,
+		Timeout: 30 * time.Second,
+	}
+}
+
+type Options struct {
+	Force   bool
+	Wait    bool
+	Timeout time.Duration
```
Comment (diafour) [inline internal/virtualization/cmd/powerstate/powerstate.go:87]: Timeout is not used. May be delete it?

Comment (yaroslavborbat) [reply internal/virtualization/cmd/powerstate/powerstate.go:87]: No, we create a context using this timeout

File: internal/virtualization/cmd/powerstate/vmop/vmop.go (hunk 2 above)
Comment (yaroslavborbat) [reply internal/virtualization/cmd/powerstate/vmop/vmop.go:117]: The method has been renamed to isFinished and is no used in the generateMsg

File: internal/virtualization/cmd/powerstate/vmop/vmop.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+		switch vmop.Spec.Type {
+		case v1alpha2.VMOPOperationTypeStart:
+			sb.WriteString("started. ")
+		case v1alpha2.VMOPOperationTypeStop:
+			sb.WriteString("stopped. ")
+		case v1alpha2.VMOPOperationTypeRestart:
+			sb.WriteString("restarted. ")
+		}
+	} else {
+		sb.WriteString("in progress. ")
```
Comment (diafour) [inline internal/virtualization/cmd/powerstate/vmop/vmop.go:84]: It is much clearer now! But "in progress" message is noisy:

> VirtualMachine "my-vm" in progress. VirtualMachineOperation "my-vm-qwd23" was not finished. phase=InProgress.

I propose to add operation type and add descriptions for Pending and InProgress phases (see next comment):

> VirtualMachine "my-vm" stopping. VirtualMachineOperation "my-vm-qwd23" is in progress.

> VirtualMachine "my-vm" starting. VirtualMachineOperation "my-vm-qwd23" pending.


```suggestion
		switch vmop.Spec.Type {
		case v1alpha2.VMOPOperationTypeStart:
			sb.WriteString("starting. ")
		case v1alpha2.VMOPOperationTypeStop:
			sb.WriteString("stopping. ")
		case v1alpha2.VMOPOperationTypeRestart:
			sb.WriteString("restarting. ")
		}
```

File: internal/virtualization/cmd/powerstate/vmop/vmop.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	sb.WriteString(fmt.Sprintf("VirtualMachineOperation %q ", key.String()))
+	switch phase {
+	case v1alpha2.VMOPPhaseCompleted:
+		sb.WriteString("completed.")
+	case v1alpha2.VMOPPhaseFailed:
+		sb.WriteString(fmt.Sprintf("failed. reason=%q, message=%q.", vmop.Status.FailureReason, vmop.Status.FailureMessage))
+	default:
+		sb.WriteString(fmt.Sprintf("was not finished. phase=%q.", phase))
```
Comment (diafour) [inline internal/virtualization/cmd/powerstate/vmop/vmop.go:94]:
```suggestion
	case v1alpha2.VMOPPhasePenging:
		sb.WriteString("pending.")
	case v1alpha2.VMOPPhaseInProgress:
		sb.WriteString("in progress.")
	case v1alpha2.VMOPPhaseCompleted:
		sb.WriteString("completed.")
	case v1alpha2.VMOPPhaseFailed:
		sb.WriteString(fmt.Sprintf("failed. reason=%q, message=%q.", vmop.Status.FailureReason, vmop.Status.FailureMessage))
	default:
		sb.WriteString(fmt.Sprintf(" phase=%q.", phase))
```

**Review by diafour — CHANGES_REQUESTED** [review]

**Review by diafour — APPROVED** [review]

File: internal/virtualization/cmd/powerstate/start.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/deckhouse/deckhouse-cli/internal/virtualization/templates"
+)
+
+func NewStartCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+	ps := NewPowerState(Start, clientConfig)
+	start := string(Start)
+	cmd := &cobra.Command{
+		Use:     start + " (VirtualMachine)",
+		Short:   strings.Title(start) + " a virtual machine.",
```
Comment (mvasl) [inline internal/virtualization/cmd/powerstate/start.go:33]: Title is deprecated
https://github.com/golang/go/issues/48367

File: internal/virtualization/cmd/powerstate/restart.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/deckhouse/deckhouse-cli/internal/virtualization/templates"
+)
+
+func NewRestartCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+	ps := NewPowerState(Restart, clientConfig)
+	restart := string(Restart)
+	cmd := &cobra.Command{
+		Use:     restart + " (VirtualMachine)",
+		Short:   strings.Title(restart) + " a virtual machine.",
```
Comment (mvasl) [inline internal/virtualization/cmd/powerstate/restart.go:33]: Title is deprecated
https://github.com/golang/go/issues/48367

File: internal/virtualization/cmd/powerstate/stop.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/deckhouse/deckhouse-cli/internal/virtualization/templates"
+)
+
+func NewStopCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+	ps := NewPowerState(Stop, clientConfig)
+	stop := string(Stop)
+	cmd := &cobra.Command{
+		Use:     stop + " (VirtualMachine)",
+		Short:   strings.Title(stop) + " a virtual machine.",
```
Comment (mvasl) [inline internal/virtualization/cmd/powerstate/stop.go:33]: Title is deprecated
https://github.com/golang/go/issues/48367

**Review by mvasl — CHANGES_REQUESTED** [review]

File: internal/virtualization/cmd/powerstate/start.go (hunk 6 above)
Comment (yaroslavborbat) [reply internal/virtualization/cmd/powerstate/start.go]: fixed

File: internal/virtualization/cmd/powerstate/stop.go (hunk 8 above)
Comment (yaroslavborbat) [reply internal/virtualization/cmd/powerstate/stop.go]: fixed

File: internal/virtualization/cmd/powerstate/restart.go (hunk 7 above)
Comment (yaroslavborbat) [reply internal/virtualization/cmd/powerstate/restart.go]: fixed

**Review by mvasl — APPROVED** [review]

**Pull request merged**